# DataFrames

So we now have a grasp of dealing with a Series object, which is a one dimensional list with a lot of built-in, convenient functions.  

Now we will learn about the next level data structure in Pandas, which is the DataFrame.  A DataFrame is a two dimensional object of data.  If a Series is a column, a DataFrame is a table. 

Hopefully, once you get the hang of things, you'll start to think in terms of DataFrames when attempting to break down your problems for analysis.  It will make your life easier. 

For reference, [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame) is the API documentation of all the funcationality of a Pandas DataFrame.  You should hold on to this site, you'll need it.

## Creating DataFrames

Remember when we created a Series from a list of values?  Now we can create a DataFrame from a multitude of ways, namely:
- a list of dictionaries
- a dictionary of lists, dictionaries, or Series objects
- a single Series
- a text file: JSON, CSV, unformatted table, Excel file
- a binary file: HDF5, Python Pickle
- an SQL query
- another DataFrame

For a full list of the possible ways to create a DataFrame, see [here](http://pandas.pydata.org/pandas-docs/stable/io.html).  If it's popular, you can probably easily turn it into a DataFrame.

We start from the top for examples to explain.

In [1]:
import pandas as pd
import numpy as np

In [4]:
# Create a list of dictionaries
employees = [
    {
        'name': 'John',
        'salary': 80000
    },
    {
        'name': 'Mary',
        'salary': 200000,
    },
    {
        'name': 'Michael',
        'salary': 100000,
    },
    {
        'name': 'Betty',
        'salary': 76000,
    },
    {
        'name': 'Greg',
        'salary': 90000,
    },
    {
        'name': 'Hillary',
        'salary': 120000,
    }    
]
# Turn it into a dataframe
df_employees = pd.DataFrame(employees)
df_employees

name  salary
0     John   80000
1     Mary  200000
2  Michael  100000
3    Betty   76000
4     Greg   90000
5  Hillary  120000

In [5]:
# How about a dictionary of lists?
employees_dict = {
    'name'  : ['John','Mary','Michael','Betty','Greg','Hillary'],
    'salary': [80000,200000,100000,76000,90000,120000],
}
df_employees = pd.DataFrame(employees_dict)
df_employees

name  salary
0     John   80000
1     Mary  200000
2  Michael  100000
3    Betty   76000
4     Greg   90000
5  Hillary  120000

The real power if creating DataFrames comes from the different file types you can create from.  The four main functions used for file I/O are:
- `from_json()`  : Create a DataFrame from JSON data, typically used when dealing with APIs over the internet
- `from_csv()`   : Create a DataFarme from a comma separated value data, a typical table-like data storage format
- `from_table()` : Create a DataFrame from textual data, highly configurable.
- `read_excel()` : Create a DataFrame from an Microsoft Excel file

In [2]:
import json
import requests

In [3]:
# Get JSON data from coinmarketcap.com's API (top 50 cryptocurrencies by market cap)
req = requests.get('https://api.coinmarketcap.com/v1/ticker/?limit=50')
res = req.text
# note that `read_json` wants a text source of data
cryptos = pd.read_json(res)
cryptos.head()
# cryptos.tail()      # Show last 5 rows
# cryptos.sample(n=5) # Show a random sample of n rows
# cryptos.describe()  # Show statistical information about DataFrame
# cryptos.info()      # Show metadata about DataFrame

24h_volume_usd  available_supply                id  last_updated  \
0      1467260000          16379562           bitcoin    1496927960   
1       452666000          92308853          ethereum    1496927968   
2       157996000       38621693933            ripple    1496927943   
3        12559500        8999999999               nem    1496927964   
4       115964000          92403434  ethereum-classic    1496927980   

   market_cap_usd              name  percent_change_1h  percent_change_24h  \
0     45065907909           Bitcoin              -0.24               -3.07   
1     23918977680          Ethereum              -0.47               -1.15   
2     11304144976            Ripple              -0.28                3.22   
3      1904247000               NEM              -1.93               -5.85   
4      1612587769  Ethereum Classic              -0.78               -2.34   

   percent_change_7d  price_btc    price_usd  rank symbol  total_supply  
0              13.57   1.000000  2751.350000     1    BTC      16379562  
1              14.03   0.094799   259.119000     2    ETH      92308853  
2              -4.11   0.000107     0.292689     3    XRP   99994661895  
3              -5.58   0.000077     0.211583     4    XEM    8999999999  
4              -0.12   0.006385    17.451600     5    ETC      92403434

In [76]:
# Reading from CSV file.
#   breaches.csv is all of the breaches in the haveibeenpwnded.com database
#   This was retrieved from their API, and turned into a CSV (using Pandas of course :) )
breaches = pd.read_csv('breaches.csv', parse_dates=['BreachDate', 'AddedDate', 'ModifiedDate'])
breaches.head()

Unnamed: 0           AddedDate BreachDate  \
0           0 2015-10-26 23:35:45 2015-03-01   
1           1 2016-10-08 07:46:05 2012-01-01   
2           2 2016-07-08 01:55:03 2016-04-19   
3           3 2016-03-06 11:07:41 2014-11-25   
4           4 2013-12-04 00:00:00 2013-10-04   

                                         DataClasses  \
0  ['Email addresses', 'IP addresses', 'Names', '...   
1                   ['Email addresses', 'Passwords']   
2  ['Device information', 'Email addresses', 'IP ...   
3  ['Dates of birth', 'Email addresses', 'IP addr...   
4  ['Email addresses', 'Password hints', 'Passwor...   

                                         Description          Domain IsActive  \
0  In approximately March 2015, the free web host...  000webhost.com     True   
1  In approximately 2012, it's alleged that the C...         126.com     True   
2  In April 2016, customer data obtained from the...        17app.co     True   
3  In November 2014, the acne website <a href="ht...        acne.org     True   
4  In October 2013, 153 million Adobe accounts we...       adobe.com     True   

  IsFabricated IsRetired IsSensitive IsSpamList IsVerified LogoType  \
0        False     False       False      False       True      png   
1        False     False       False      False      False      svg   
2        False     False       False      False       True      svg   
3        False     False       False      False       True      svg   
4        False     False       False      False       True      svg   

         ModifiedDate        Name   PwnCount       Title  
0 2015-10-26 23:35:45  000webhost   13545468  000webhost  
1 2016-10-08 07:46:05         126    6414191         126  
2 2016-07-08 01:55:03     17Media    4009640          17  
3 2016-03-06 11:07:41     AcneOrg     432943    Acne.org  
4 2013-12-04 00:00:00       Adobe  152445165       Adobe

and one more for good measure:

In [5]:
cars = pd.read_csv('../Day2_old/Day 2 - For Corey/data/mtcars.csv')
cars.head()

Model   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0           Mazda X4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

## Data Slicing and Selection

Getting the original data into a DataFrame is just the first step.  If you didn't need to do anything to the data, like perform analysis, look for relatonships, or get a subset of it (typically called a slice), then there is no reason to import it at all.  Let's take a look at the basic operations of "looking" at the DataFrame.

All of the methods described here **will always give you less information than the original DataFrame**, with no alteration of the data.  Later we will look at how to add new columns and change the information within.

As always, [here](http://pandas.pydata.org/pandas-docs/stable/indexing.html) is the documentation for indexing and slicing.

It is also worth mentioning that the methods used for accessing information in Python dictionaries will typically work, but in some cases it may fail.  Therefore, it is advised to use the specialized selection mechanisms in Pandas, especially when dealing with slicing.

#### Getting a row, column, or a value

Let's just return one of the columns of a DataFrame.  Let's use the cryptocurrency data from earlier, because I love cryptocurrencies.  In order to get a row, you can just great it like you would a dictionary, and call the name of the column you want in quotes.

In [50]:
cryptos['name']

0              Bitcoin
1               Ripple
2             Ethereum
3             Litecoin
4                  NEM
5                 Dash
6     Ethereum Classic
7       Stellar Lumens
8               Monero
9             Bytecoin
10               Steem
11               Golem
12               Augur
13        MaidSafeCoin
14               Waves
15              Gnosis
16         GameCredits
17            Dogecoin
18               Round
19             Stratis
20               Zcash
21           BitShares
22             Siacoin
23              Decred
24               Ardor
25            DigixDAO
26         SingularDTV
27              Factom
28             Iconomi
29          BitConnect
30              Tether
31                Lisk
32            DigiByte
33                 Nxt
34                PIVX
35            Byteball
36            Peercoin
37           iExec RLC
38          FirstBlood
39            Emercoin
40         Storjcoin X
41             SysCoin
42              Komodo
43         

Equivalently, if your column names don't have spaces or numbers, you can call a column by doing the `dot` method.

In [55]:
cryptos.name  # Ready out loud "cryptos dot name"

0              Bitcoin
1               Ripple
2             Ethereum
3             Litecoin
4                  NEM
5                 Dash
6     Ethereum Classic
7       Stellar Lumens
8               Monero
9             Bytecoin
10               Steem
11               Golem
12               Augur
13        MaidSafeCoin
14               Waves
15              Gnosis
16         GameCredits
17            Dogecoin
18               Round
19             Stratis
20               Zcash
21           BitShares
22             Siacoin
23              Decred
24               Ardor
25            DigixDAO
26         SingularDTV
27              Factom
28             Iconomi
29          BitConnect
30              Tether
31                Lisk
32            DigiByte
33                 Nxt
34                PIVX
35            Byteball
36            Peercoin
37           iExec RLC
38          FirstBlood
39            Emercoin
40         Storjcoin X
41             SysCoin
42              Komodo
43         

What we get is a Series object in return.  This is why it was necessary to discuss Series beforehand, because a good portion of the return results when working with Pandas will be a one dimensional Series object.

#### Getting Multiple Columns

You can also pass a list of column names to return a new DataFrame instead.  This will more often be the case when working with Pandas.  It is obvious you can't use the `dot` method for multiple columns. 

In [51]:
cryptos[['name', 'market_cap_usd']]

name  market_cap_usd
0            Bitcoin     29959243064
1             Ripple     14028798491
2           Ethereum      8294221249
3           Litecoin      1284033909
4                NEM      1070127000
5               Dash       618777832
6   Ethereum Classic       545087952
7     Stellar Lumens       423320448
8             Monero       385667941
9           Bytecoin       201378070
10             Steem       193382225
11             Golem       166081034
12             Augur       160527400
13      MaidSafeCoin       141303155
14             Waves       139966000
15            Gnosis       134532434
16       GameCredits       132194217
17          Dogecoin       131353261
18             Round       129772900
19           Stratis       123770172
20             Zcash       117801879
21         BitShares       114281589
22           Siacoin        94924542
23            Decred        89871066
24             Ardor        88871794
25          DigixDAO        88857600
26       SingularDTV        86959800
27            Factom        74052933
28           Iconomi        65672472
29        BitConnect        64165045
30            Tether        59406728
31              Lisk        56993199
32          DigiByte        56796244
33               Nxt        56675667
34              PIVX        55504983
35          Byteball        41550081
36          Peercoin        41467246
37         iExec RLC        37994088
38        FirstBlood        37577236
39          Emercoin        36367571
40       Storjcoin X        33948366
41           SysCoin        33755003
42            Komodo        33233987
43             Lykke        25748965
44         AntShares        25659100
45          Namecoin        25145604
46             Nexus        22143375
47       BitcoinDark        21765531
48            Xaurum        21641992
49      Counterparty        21481350

#### What About Rows?

In order to just get a certain row, you need to call the index of that row.  Remember the default index is the range of the length of the DataFrame, so we would need to call the number that corresponds to "how far" that row is down the table.

There are two ways to call rows: `loc` and `iloc`.  The former `loc` is strictly for calling label-like indices, where as the latter `iloc` is the integer-based index.  Note that if we are using the default index, it is the same thing.  But in the case of text as our index, we can directly call a certain row by its label.

In [58]:
cryptos.loc[0]

24h_volume_usd        1.05738e+09
available_supply         16335287
id                        bitcoin
last_updated           1495030151
market_cap_usd        29959243064
name                      Bitcoin
percent_change_1h            0.32
percent_change_24h           5.21
percent_change_7d            4.32
price_btc                       1
price_usd                 1834.02
rank                            1
symbol                        BTC
total_supply             16335287
Name: 0, dtype: object

In [56]:
cryptos.iloc[0]

24h_volume_usd        1.05738e+09
available_supply         16335287
id                        bitcoin
last_updated           1495030151
market_cap_usd        29959243064
name                      Bitcoin
percent_change_1h            0.32
percent_change_24h           5.21
percent_change_7d            4.32
price_btc                       1
price_usd                 1834.02
rank                            1
symbol                        BTC
total_supply             16335287
Name: 0, dtype: object

#### Boolean Mask

So we can slice and dice our dataset to our hearts content, but what if our data is large, and we want to constrain it with certain restrictions?  That's where the Boolean Mask comes into play, and it is awesome.

Let's see what we get when we just make a simple logical comparison with a column of a DataFrame

In [61]:
cryptos.percent_change_7d > 25

0     False
1      True
2     False
3     False
4     False
5     False
6     False
7      True
8     False
9      True
10    False
11    False
12    False
13    False
14    False
15    False
16     True
17    False
18     True
19    False
20    False
21    False
22     True
23    False
24     True
25    False
26    False
27    False
28     True
29     True
30    False
31    False
32     True
33     True
34    False
35    False
36    False
37    False
38    False
39    False
40     True
41    False
42    False
43    False
44    False
45    False
46    False
47     True
48    False
49    False
Name: percent_change_7d, dtype: bool

We get a Series object of True/False values which correspond to whether or not that row's value makes the queried logic True or False.  This is called a "Boolean Mask" and if we pass the entire mask into the original DataFrame, we will have sliced it to return only the values that satisfy our initial request.

In [62]:
cryptos[cryptos.percent_change_7d > 25]

24h_volume_usd  available_supply            id  last_updated  \
1      478549000.0       38305873865        ripple    1495030143   
7       68127900.0        9557384381       stellar    1495030156   
9       15340800.0      182899712054  bytecoin-bcn    1495030153   
16       7057430.0          62705375   gamecredits    1495030157   
18       1025000.0         850000000         round    1495030183   
22      15736400.0       26022051924       siacoin    1495030167   
24       1267950.0         998999495         ardor    1495030178   
28        904287.0          87000000       iconomi    1495030181   
29        845224.0           6382232    bitconnect    1495029887   
32      66367100.0        8107031908      digibyte    1495030147   
33       4160490.0         998999983           nxt    1495030144   
40       2225840.0          51173144   storjcoin-x    1495030158   
47        500283.0           1288862   bitcoindark    1495030155   

    market_cap_usd            name  percent_change_1h  percent_change_24h  \
1      14028798491          Ripple               7.18               10.61   
7        423320448  Stellar Lumens              11.90              -10.82   
9        201378070        Bytecoin              -2.37              133.85   
16       132194217     GameCredits              -3.15               -7.35   
18       129772900           Round               1.65               -8.29   
22        94924542         Siacoin               4.41               42.27   
24        88871794           Ardor               2.43               -1.65   
28        65672472         Iconomi               1.05               -2.35   
29        64165045      BitConnect               1.04                3.49   
32        56796244        DigiByte               6.14              177.43   
33        56675667             Nxt              -1.62               -6.06   
40        33948366     Storjcoin X               5.53               33.56   
47        21765531     BitcoinDark               1.08                6.22   

    percent_change_7d     price_btc  price_usd  rank symbol  total_supply  
1              110.49  2.001500e-04   0.366231     2    XRP   99994799156  
7               43.63  2.421000e-05   0.044292     8    XLM  102901180739  
9              107.94  6.000000e-07   0.001101    10    BCN  182899712054  
16             114.28  1.152130e-03   2.108180    17   GAME      62705375  
18             288.94  8.344000e-05   0.152674    19  ROUND    1000000000  
22              55.43  1.990000e-06   0.003648    23     SC   26022051924  
24              35.35  4.862000e-05   0.088961    25   ARDR     998999495  
28              26.06  4.125300e-04   0.754856    29    ICN     100000000  
29              30.44  5.489820e-03  10.053700    30    BCC       6382232  
32             285.12  3.830000e-06   0.007006    33    DGB    8107031908  
33              30.11  3.100000e-05   0.056732    34    NXT     998999983  
40              38.13  3.625500e-04   0.663402    41   SJCX     500000000  
47              35.40  9.229050e-03  16.887400    48   BTCD       1288862

Now this is a powerful way of querying our DataFrame to get the data we want.  What is nice is that you can string together arbitrary amounts of logic together to further constrain the returned data.  You just have to wrap each query in parenthesis and stick an `&` in the middle of them. 

In [66]:
cryptos[
    (cryptos.percent_change_7d > 25) &
    (cryptos.market_cap_usd > 200000000)
]

24h_volume_usd  available_supply            id  last_updated  \
1     478549000.0       38305873865        ripple    1495030143   
7      68127900.0        9557384381       stellar    1495030156   
9      15340800.0      182899712054  bytecoin-bcn    1495030153   

   market_cap_usd            name  percent_change_1h  percent_change_24h  \
1     14028798491          Ripple               7.18               10.61   
7       423320448  Stellar Lumens              11.90              -10.82   
9       201378070        Bytecoin              -2.37              133.85   

   percent_change_7d     price_btc  price_usd  rank symbol  total_supply  
1             110.49  2.001500e-04   0.366231     2    XRP   99994799156  
7              43.63  2.421000e-05   0.044292     8    XLM  102901180739  
9             107.94  6.000000e-07   0.001101    10    BCN  182899712054

There's also the `query` method built-in as well, which does the same thing as Boolean Mask, but in somewhat of a more shorthand method.  

`query` uses a boolean string to select information

Let's look at it.  

In [21]:
cars.query('mpg > 15')             # Simple query
# cars.query('20 > mpg >= 15')       # more complex query
# cars.query('mpg > 20 and cyl < 6') # multiple booleans
# cars.query('index == mpg')         # Use index in query ***

Model   mpg  cyl   disp  hp  drat    wt  qsec  vs  am  gear  carb
26  Porsche 914-2  26.0    4  120.3  91  4.43  2.14  16.7   0   1     5     2

### Selecting Strings
You can also check to see if something contains a certain string when doing selections.
* Select only Toyotas

In [64]:
cars[cars.Model.str.contains('Toyota')]

Model   mpg  cyl   disp  hp  drat     wt   qsec  vs  am  gear  \
19  Toyota Corolla  33.9    4   71.1  65  4.22  1.835  19.90   1   1     4   
20   Toyota Corona  21.5    4  120.1  97  3.70  2.465  20.01   1   0     3   

    carb  
19     1  
20     1

Or use regex!
* Select cars with 2 numbers in the name 

In [63]:
cars[cars.Model.str.contains(r'[0-9][0-9]+')]

Model   mpg  cyl   disp   hp  drat    wt   qsec  vs  am  gear  \
2      Datsun 710  22.8    4  108.0   93  3.85  2.32  18.61   1   1     4   
6      Duster 360  14.3    8  360.0  245  3.21  3.57  15.84   0   0     3   
7       Merc 240D  24.4    4  146.7   62  3.69  3.19  20.00   1   0     4   
8        Merc 230  22.8    4  140.8   95  3.92  3.15  22.90   1   0     4   
9        Merc 280  19.2    6  167.6  123  3.92  3.44  18.30   1   0     4   
10      Merc 280C  17.8    6  167.6  123  3.92  3.44  18.90   1   0     4   
11     Merc 450SE  16.4    8  275.8  180  3.07  4.07  17.40   0   0     3   
12     Merc 450SL  17.3    8  275.8  180  3.07  3.73  17.60   0   0     3   
13    Merc 450SLC  15.2    8  275.8  180  3.07  3.78  18.00   0   0     3   
17       Fiat 128  32.4    4   78.7   66  4.08  2.20  19.47   1   1     4   
23     Camaro Z28  13.3    8  350.0  245  3.73  3.84  15.41   0   0     3   
26  Porsche 914-2  26.0    4  120.3   91  4.43  2.14  16.70   0   1     5   

    carb  
2      1  
6      4  
7      2  
8      2  
9      4  
10     4  
11     3  
12     3  
13     3  
17     1  
23     4  
26     2

## Manipulating DataFrames

Almost all method of manipulation have the `inplace` option, which gives you the option to return a new DataFrame, or replace the existing DataFrame in memory. 

I personally tend to leave the original DataFrames in place, un-edited, and create new DataFrames that can be used.  `inplace` is a good method for initial curation of imported data.  Once you have it properly "munged," not altering it further is good practice.  

### Setting the index

The index is what is optimized for searching when dealing with DataFrames.  For large datasets, it is often useful to set the index on something you'd like to search on most.

This is done through the `set_index()` method.

In [71]:
breaches.set_index('Name', inplace=True)
breaches.head()

Unnamed: 0             AddedDate  BreachDate  \
Name                                                       
000webhost           0  2015-10-26T23:35:45Z  2015-03-01   
126                  1  2016-10-08T07:46:05Z  2012-01-01   
17Media              2  2016-07-08T01:55:03Z  2016-04-19   
AcneOrg              3  2016-03-06T11:07:41Z  2014-11-25   
Adobe                4  2013-12-04T00:00:00Z  2013-10-04   

                                                  DataClasses  \
Name                                                            
000webhost  ['Email addresses', 'IP addresses', 'Names', '...   
126                          ['Email addresses', 'Passwords']   
17Media     ['Device information', 'Email addresses', 'IP ...   
AcneOrg     ['Dates of birth', 'Email addresses', 'IP addr...   
Adobe       ['Email addresses', 'Password hints', 'Passwor...   

                                                  Description          Domain  \
Name                                                                            
000webhost  In approximately March 2015, the free web host...  000webhost.com   
126         In approximately 2012, it's alleged that the C...         126.com   
17Media     In April 2016, customer data obtained from the...        17app.co   
AcneOrg     In November 2014, the acne website <a href="ht...        acne.org   
Adobe       In October 2013, 153 million Adobe accounts we...       adobe.com   

           IsActive IsFabricated IsRetired IsSensitive IsSpamList IsVerified  \
Name                                                                           
000webhost     True        False     False       False      False       True   
126            True        False     False       False      False      False   
17Media        True        False     False       False      False       True   
AcneOrg        True        False     False       False      False       True   
Adobe          True        False     False       False      False       True   

           LogoType          ModifiedDate   PwnCount       Title  
Name                                                              
000webhost      png  2015-10-26T23:35:45Z   13545468  000webhost  
126             svg  2016-10-08T07:46:05Z    6414191         126  
17Media         svg  2016-07-08T01:55:03Z    4009640          17  
AcneOrg         svg  2016-03-06T11:07:41Z     432943    Acne.org  
Adobe           svg  2013-12-04T00:00:00Z  152445165       Adobe

### Deleting Data

What if you have columns you don't want?  The `del` command will get rid of it for you.  Be careful though!  It does not return a new DataFrame like the other methods we've looked at, it alters the current DataFrame in memory.

Basically it is 
```python
inplace=True
```
by default

In [13]:
del(breaches['Unnamed: 0'])  # Deleting the index row of the previously imported DataFrame, because its redundant

In [68]:
breaches

AddedDate  BreachDate  \
0    2015-10-26T23:35:45Z  2015-03-01   
1    2016-10-08T07:46:05Z  2012-01-01   
2    2016-07-08T01:55:03Z  2016-04-19   
3    2016-03-06T11:07:41Z  2014-11-25   
4    2013-12-04T00:00:00Z  2013-10-04   
5    2015-05-22T06:03:44Z  2015-05-21   
6    2014-11-06T21:47:52Z  2013-05-30   
7    2016-11-07T21:55:29Z  2016-09-27   
8    2015-12-20T06:47:19Z  2011-10-30   
9    2017-05-04T22:07:38Z  2016-12-16   
10   2016-11-10T03:24:38Z  2016-05-18   
11   2015-08-18T20:55:05Z  2015-07-19   
12   2014-07-06T03:49:45Z  2013-12-19   
13   2016-10-01T23:42:56Z  2015-12-06   
14   2016-03-12T22:08:58Z  2014-05-26   
15   2016-07-06T08:16:03Z  2013-06-01   
16   2014-01-23T13:10:29Z  2011-06-26   
17   2016-04-25T10:05:34Z  2015-11-11   
18   2014-02-01T23:57:10Z  2014-02-01   
19   2017-05-16T01:49:31Z  2017-05-15   
20   2014-04-08T05:44:10Z  2014-04-03   
21   2014-09-10T20:30:11Z  2014-01-09   
22   2017-03-27T23:45:41Z  2015-05-22   
23   2016-06-08T10:49:24Z  2016-01-01   
24   2015-11-03T22:20:17Z  2014-06-23   
25   2016-12-27T08:24:52Z  2014-11-13   
26   2014-03-30T13:07:16Z  2014-03-29   
27   2016-09-05T10:02:23Z  2013-04-01   
28   2017-03-12T03:21:52Z  2014-10-01   
29   2014-05-11T04:25:48Z  2014-04-25   
..                    ...         ...   
185  2016-05-29T22:59:04Z  2013-02-28   
186  2016-06-27T08:07:18Z  2016-06-01   
187  2014-02-23T04:32:08Z  2014-02-20   
188  2016-11-07T09:04:54Z  2016-08-11   
189  2016-11-05T22:32:39Z  2016-01-14   
190  2016-12-27T10:05:41Z  2016-09-06   
191  2016-01-24T13:15:11Z  2015-11-03   
192  2014-07-06T04:16:37Z  2014-01-10   
193  2016-06-09T09:16:36Z  2012-01-01   
194  2013-11-30T00:00:00Z  2013-12-12   
195  2015-11-25T07:00:57Z  2015-11-13   
196  2016-11-07T11:07:25Z  2012-07-04   
197  2016-07-21T02:25:49Z  2014-11-24   
198  2016-06-28T23:47:07Z  2012-05-21   
199  2016-09-06T05:51:12Z  2015-09-25   
200  2016-03-06T21:41:16Z  2015-07-11   
201  2014-06-01T10:01:32Z  2013-09-03   
202  2017-03-15T19:29:52Z  2016-08-07   
203  2014-02-01T02:57:21Z  2014-01-04   
204  2016-08-05T06:53:35Z  2015-11-04   
205  2017-01-29T07:20:52Z  2015-09-25   
206  2016-02-07T20:26:56Z  2015-02-01   
207  2015-08-08T04:28:48Z  2013-11-07   
208  2013-12-04T00:00:00Z  2012-07-11   
209  2014-09-12T04:50:32Z  2014-09-07   
210  2017-04-15T11:02:35Z  2016-12-01   
211  2015-07-30T05:32:00Z  2012-02-21   
212  2017-02-08T07:59:39Z  2011-01-01   
213  2016-12-28T07:03:17Z  2016-08-08   
214  2015-05-12T23:50:08Z  2015-05-11   

                                           DataClasses  \
0    ['Email addresses', 'IP addresses', 'Names', '...   
1                     ['Email addresses', 'Passwords']   
2    ['Device information', 'Email addresses', 'IP ...   
3    ['Dates of birth', 'Email addresses', 'IP addr...   
4    ['Email addresses', 'Password hints', 'Passwor...   
5    ['Dates of birth', 'Email addresses', 'Genders...   
6    ['Email addresses', 'Genders', 'Geographic loc...   
7                     ['Email addresses', 'Passwords']   
8    ['Dates of birth', 'Email addresses', 'Homepag...   
9                     ['Email addresses', 'Passwords']   
10   ['Avatars', 'Email addresses', 'Geographic loc...   
11   ['Dates of birth', 'Email addresses', 'Ethnici...   
12   ['Email addresses', 'Instant messenger identit...   
13   ['Email addresses', 'IP addresses', 'Passwords...   
14       ['Email addresses', 'Passwords', 'Usernames']   
15   ['Dates of birth', 'Email addresses', 'Genders...   
16                          ['Passwords', 'Usernames']   
17   ['Beauty ratings', 'Car ownership statuses', '...   
18   ['Credit cards', 'Genders', 'Passwords', 'User...   
19   ['Email addresses', 'Geographic locations', 'I...   
20   ['Career levels', 'Education levels', 'Email a...   
21                    ['Email addresses', 'Passwords']   
22   ['Dates of birth', 'Email addresses', 'Genders...   
23   ['Email addresses', 'IP addresses', 'Passwords...   
24   ['Dates of birth',

### Adding Columns

Up until now, we've only discussed how to query and get information from the data that is already in your DataFrame.  This is great, but often we want to combine columns, or produce data _from_ the data and add it onto the DataFrame.  Pandas makes this as easy as pie.  

Let's start by doing some math on a column, to produce a more easily readable number of crypto market caps.  

Note that like the del function, we are altering the current DataFrame in memory, not creating a new one.

In [70]:
# adding the "market_cap_usd_billions" column to the dataframe by manipulating the "market_cap_usd" column
cryptos['market_cap_usd_billions'] = cryptos.market_cap_usd / 1e9  
cryptos.head()

24h_volume_usd  available_supply        id  last_updated  market_cap_usd  \
0    1.057380e+09          16335287   bitcoin    1495030151     29959243064   
1    4.785490e+08       38305873865    ripple    1495030143     14028798491   
2    1.902740e+08          91685426  ethereum    1495030166      8294221249   
3    1.371880e+08          51144707  litecoin    1495030141      1284033909   
4    1.563740e+07        8999999999       nem    1495030163      1070127000   

       name  percent_change_1h  percent_change_24h  percent_change_7d  \
0   Bitcoin               0.32                5.21               4.32   
1    Ripple               7.18               10.61             110.49   
2  Ethereum               0.48               -0.29               1.50   
3  Litecoin              -1.63                4.14             -27.48   
4       NEM               1.84                1.94               8.70   

   price_btc    price_usd  rank symbol  total_supply  market_cap_usd_billions  
0   1.000000  1834.020000     1    BTC      16335287                29.959243  
1   0.000200     0.366231     2    XRP   99994799156                14.028798  
2   0.049439    90.463900     3    ETH      91685426                 8.294221  
3   0.013721    25.105900     4    LTC      51144707                 1.284034  
4   0.000065     0.118903     5    XEM    8999999999                 1.070127

Easy as that.  When dealing with scalars, you act like you're working on Series object (you are actually) and you "tack" on the resulting Series. You can do the same by combining columns together, as long as the resulting object is as long as the length of the DataFrame.

## Getting a quick histogram

When you'd like to know how many of a certain class there are in a Series, you can quickly get the number of times they appear with `value_counts`.

In [67]:
breaches.DataClasses.value_counts()

['Email addresses', 'IP addresses', 'Passwords', 'Usernames']                                                                                                                                                                                                                                                                                                                                                                                                 30
['Email addresses', 'Passwords', 'Usernames']                                                                                                                                                                                                                                                                                                                                                                                                                 28
['Email addresses', 'Passwords']                                                                      

### map, apply, applymap

Often times you'll want to perform something more complicated, like a function to the entire row, each element in a row, or to the entire DataFrame.  Guess what, Pandas has something for that.

It turns out, that a lot of the built-in functions we saw in the Series section are actually just specifications of the `apply` method.  `apply` allows you to apply functions to the entire row/column

In [15]:
# Turn everything into relative percentages of the whole column (i.e. normalization)
perc = lambda x: x/x.max() * 100
perc_df = cryptos.set_index('name')[['24h_volume_usd', 'market_cap_usd', 'price_btc']].apply(perc)
perc_df.head()

24h_volume_usd  market_cap_usd   price_btc
name                                                        
Bitcoin               100.000000      100.000000  100.000000
Ethereum               34.304635       51.188099    9.064160
Ripple                  6.414104       23.503287    0.009942
NEM                     0.705213        4.441812    0.008063
Ethereum Classic        4.410764        3.417593    0.604599

Similarly, `applymap` will do so in an element-wise fashion, instead of taking the whole row into account.  This is usually used for formatting purposes, maybe outputting only two decimals of a floating point number (canonical example)

In [18]:
two_digits = lambda x: '%.2f' % x
perc_df.applymap(two_digits).head()

24h_volume_usd market_cap_usd price_btc
name                                                    
Bitcoin                  100.00         100.00    100.00
Ethereum                  34.30          51.19      9.06
Ripple                     6.41          23.50      0.01
NEM                        0.71           4.44      0.01
Ethereum Classic           4.41           3.42      0.60

`map` will do the same as `applymap`, but works on Series objects, not DataFrames.

In [23]:
perc_df.map(two_digits)
#perc_df['24h_volume_usd'].map(two_digits)

AttributeError: 'DataFrame' object has no attribute 'map'

Summing up, apply works on a row / column basis of a DataFrame, applymap works element-wise on a DataFrame, and map works element-wise on a Series.

### Merging Data

Many times you have mulitiple data sources that describe the same things, and you'd like to combine them into a single table for comparisons.  Pandas makes this easy by giving you all types of SQL like operations for joining, merging, and concatenating data together.

The graphic given in the beginning of the class can help you chose which function to use when you want to join two sets of data in a certain way.  I use it often. 

* The primary function for joining is pandas.merge
```python
merge(left, right, how='left', on=None, left_on=None, right_on=None)
```
  * left and right specify the dataframes to join
  * how specifies the type of join to use
  * on specifies a column name present in both frames
  * left_on and right_on can specify columns that don't use the same name

## Joining Data Frames Example

In [56]:
cool = [ [cars[1].Model, "Cool"] 
        if cars[1].hp >= 200 else [cars[1].Model, "Boring"] 
        for cars in cars.iterrows()]

coolCars = pd.DataFrame(cool, columns=['Model', 'Rating'])

In [57]:
#a new data frame

coolCars[5:10]

Model  Rating
5     Valiant  Boring
6  Duster 360    Cool
7   Merc 240D  Boring
8    Merc 230  Boring
9    Merc 280  Boring

In [59]:
#To Join the two dataframes, just call merge

ratedCars = pd.merge(cars, 
                     coolCars, 
                     on=['Model'])

ratedCars[5:10]

Model   mpg  cyl   disp   hp  drat    wt   qsec  vs  am  gear  carb  \
5     Valiant  18.1    6  225.0  105  2.76  3.46  20.22   1   0     3     1   
6  Duster 360  14.3    8  360.0  245  3.21  3.57  15.84   0   0     3     4   
7   Merc 240D  24.4    4  146.7   62  3.69  3.19  20.00   1   0     4     2   
8    Merc 230  22.8    4  140.8   95  3.92  3.15  22.90   1   0     4     2   
9    Merc 280  19.2    6  167.6  123  3.92  3.44  18.30   1   0     4     4   

   Rating  
5  Boring  
6    Cool  
7  Boring  
8  Boring  
9  Boring

### Resampling Time-series data

Pandas also comes with a extensive package with working with timeseries data.  Documentation [here](https://pandas.pydata.org/pandas-docs/stable/timeseries.html).  Most of the time, I take advantage of the `resample` method, to group data into regular intervals for comparison.

In [98]:
breaches.set_index('BreachDate').resample('3M').sum().PwnCount

BreachDate
2007-07-31      2136520.0
2007-10-31            NaN
2008-01-31            NaN
2008-04-30       252216.0
2008-07-31    359420698.0
2008-10-31            NaN
2009-01-31      5774783.0
2009-04-30            NaN
2009-07-31            NaN
2009-10-31            NaN
2010-01-31            NaN
2010-04-30            NaN
2010-07-31      3619948.0
2010-10-31       590954.0
2011-01-31     56465651.0
2011-04-30     40767652.0
2011-07-31     34773100.0
2011-10-31       745355.0
2012-01-31    159529431.0
2012-04-30     46788853.0
2012-07-31    234867214.0
2012-10-31      1074948.0
2013-01-31      8269070.0
2013-04-30     67840955.0
2013-07-31    145069292.0
2013-10-31    204201162.0
2014-01-31     13612422.0
2014-04-30     94306259.0
2014-07-31      2273233.0
2014-10-31      7547880.0
2015-01-31      2058034.0
2015-04-30     50609935.0
2015-07-31     48052277.0
2015-10-31    283566853.0
2016-01-31     61551373.0
2016-04-30     40031380.0
2016-07-31     44497354.0
2016-10-31    777459346.0
2

### Exercise

* Write a script to join the following files together into a single data frame
  * state_x77.csv: a csv with various state names and data
  * state_area.csv: a csv with state names, a division factor, and the area
  * state_divs.csv: a single series which matches the state area ordering and contains the region of each state

### Grouping Data

Lots of data is categorical, meaning that values belong to some set of a group, but the statistical values associated with the groups is hard to get at.  Pandas makes it easy to to group all values of a given category together, and then aggregate appropriately to get the statistical information you need.

The `groupby` function allows us to group data into buckets of categorical data, such as male/female, True/False, etc.

Let's try it on the "breaches" dataset.

In [5]:
breaches.groupby('IsVerified')

Well what is that?  When you use the `groupby` function, it doesn't return the data, it returns the instructions on how to do the grouping.  In order to get data from it, you need to provide futher instruction on how to aggregate the groups together.  This grouping, we will want to know the size of each group, so we use the `size` function afterwards.

In [6]:
breaches.groupby('IsVerified').size()

IsVerified
False     16
True     199
dtype: int64

This tells us that there are 16 'IsVerified' rows that give False and 199 that give True.  Now we could have gotten to this same information through methodology we've already learned before.  What's special about `groupby` is its ability to handle multiple levels of groups, or groups within groups. 

Let's change `size` to `sum` and instead pass a list of columns to see what it gives us.

In [54]:
breaches.groupby(['IsVerified', 'IsSensitive']).sum()

Unnamed: 0  IsActive  IsFabricated  IsRetired  \
IsVerified IsSensitive                                                  
False      False               808      12.0           0.0        0.0   
           True                366       4.0           2.0        0.0   
True       False             20192     181.0           0.0        1.0   
           True               1639      18.0           0.0        0.0   

                        IsSpamList    PwnCount  
IsVerified IsSensitive                          
False      False               0.0  1376689242  
           True                0.0   190405201  
True       False               4.0  2073476627  
           True                0.0   111776152

Notice that when we used `sum` we didn't get any textual data, and we were returned a new DataFrame instead of a Series?  That's because Pandas only returns what is useful for the `sum` function, which is numerical data.  Also, if there are multiple columns that it can apply `sum` to, it will return the appropriate data structure. 

Now we have further subgrouping under the initial grouping. The table above allows you to answer at least 20 questions (4 rows x 5 columns).  For instance, there are 111776152 instances where someones information was leaked from a breach that is both `Verified` and `Sensitive`.  

The appropriate analysis changes with the data, but being able to quickly group categorical data allows you to quickly get answers to questions about subgroups of your data.

### Pivoting Data

Those of you who have worked a lot with Excel Spreadsheets will have probably heard of a pivot table.  Pandas can also pivot data, so that you can display the data within a table however you like.  Pandas also allows for multiple dimensions in the indices. 

Pivot tables work well when you want to compare against some of your columns that are classifications like: Male/Female, small/medium/large, etc.

API documentation for `pivot_table` can be found [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.pivot_table.html) but the general idea is:
  * ```index=``` indicates rows
  * ```column=``` indicates the columns to break out
  * ```aggfunc=``` specifies an aggregation function

In [51]:
cars.pivot_table(index=['cyl'], columns=['am'], aggfunc=np.mean)

mpg                   disp                 hp                  drat  \
am        0          1           0      1           0           1         0   
cyl                                                                           
4    22.900  29.028571  135.866667   89.7   84.666667   78.000000  3.770000   
6    19.125  20.566667  204.550000  155.0  115.250000  131.666667  3.420000   
8    15.050  15.400000  357.616667  326.0  194.166667  299.500000  3.120833   

                     wt               qsec              vs            \
am          1         0         1        0          1    0         1   
cyl                                                                    
4    4.194286  2.935000  1.936857  20.9700  18.428571  1.0  0.857143   
6    3.806667  3.388750  2.755000  19.2150  16.326667  1.0  0.000000   
8    3.880000  4.104083  3.370000  17.1425  14.550000  0.0  0.000000   

         gear                carb            
am          0         1         0         1  
cyl                                          
4    3.666667  4.285714  1.666667  1.428571  
6    3.500000  4.333333  2.500000  4.666667  
8    3.000000  5.000000  3.083333  6.000000

The `pivot` command is similar in function, but is made for dealing with non-numerical values.  See the documentation [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.pivot.html#pandas.DataFrame.pivot).

## Exporting DataFrames

Once you've done your analysis, it's often prudent to export the new dataset into a file, either for record keeping or sharing purposes.  Pandas also has many built-in routines that allow you to export the data in whatever format you or your colleagues are comfortable with. Let's go through a few.

[Here](http://pandas.pydata.org/pandas-docs/stable/io.html) is a list of all the I/O tools available.  I tend to use the following two the most.

In [ ]:
cryptos.to_json()
cryptos.to_csv()

## Exercises

Using the `requests` package and your sweet new skills manipulating DataFrames, answer the following questions:

1. Retrieve the JSON data from this site (https://api.chucknorris.io/jokes/search?query=fist) and put it into a dataframe named `chuck`

2.) Find the following information:
- How big is the DataFrame
- What are the Column names?
- Is there a good index column?  If so, what is it?

3.) Find the following information:
- What is the icon_url value of the 26th row
- Return a list of the jokes from the last 10 rows.

4.) Add a column named `length` to the DataFrame that gives the character length of the jokes. 
- What is the average length of the jokes in your DataFrame?
- What is the longest and shortest joke? 
- Using `applymap`, create a new column that gives the difference in length from the longest joke. 

4.) Enter another query instead of "fist," and place that response into a new DataFrame named what your query is.  For example, if you replace "fist" with "dead," then name your new DataFrame `dead`. Then:
- Concatenate the two datasets by rows.
- Merge the two datasets on a proper column, both inner and outer

Are there any overlaps between the two sets?

How much did your dataset grow by?

5.) Export the resulting DataFrame from problem 4 into your format of choice.  The exported dataset should have a proper index column, not just a list of numbers.